src/tests/e2etest/ZtsiTests.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using NUnit.Framework;
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.Azure.Devices;
+using System.Text.Json;
+using System;
+using System.Threading.Tasks;
+using System.Text.RegularExpressions;
+
+namespace e2etesting
+{
+    [TestFixture]
+    public class ZtsiTests : E2eTest
+    {   
+        string ComponentName = "Ztsi";
+        public partial class Ztsi
+        {
+            public int Enabled { get; set; }
+            public string ServiceUrl { get; set; }
+        }
+        public partial class DesiredZtsi
+        {
+            public bool DesiredEnabled { get; set; }
+            public string DesiredServiceUrl { get; set; }
+        }
+
+        public partial class ResponseCode
+        {
+            public int ac { get; set; }
+        }
+        private int m_twinTimeoutSeconds;
+
+        [SetUp]
+        public void TestSetUp()
+        {
+            m_twinTimeoutSeconds = base.twinTimeoutSeconds;
+            base.twinTimeoutSeconds = base.twinTimeoutSeconds * 2;
+        }
+
+        [TearDown]
+        public void TestTearDown()
+        {
+            base.twinTimeoutSeconds = m_twinTimeoutSeconds;
+        }
+
+        [Test]      
+        public void ZtsiTest_Get()
+        {
+            if ((GetTwin().ConnectionState == DeviceConnectionState.Disconnected) || (GetTwin().Status == DeviceStatus.Disabled))
+            {
+                Assert.Fail("Module is disconnected or is disabled");
+            }
+
+            Ztsi deserializedReportedObject = JsonSerializer.Deserialize<Ztsi>(GetTwin().Properties.Reported[ComponentName].ToString());
+            Assert.True(IsRegexMatch(new Regex(@"[0-2]"), deserializedReportedObject.Enabled));
+            Assert.True(deserializedReportedObject.ServiceUrl is string);
+        }
+
+        [Test]
+        public void ZtsiTest_Set_Get()
+        {
+            if ((GetNewTwin().ConnectionState == DeviceConnectionState.Disconnected) || (GetTwin().Status == DeviceStatus.Disabled))
+            {
+                Assert.Fail("Module is disconnected or is disabled");
+            }
+
+            // Test ServiceUrl
+            var desiredZtsi = new DesiredZtsi
+            {
+                DesiredServiceUrl = "https://ztsiendpoint.com"
+            };
+            var expectedZtsi = new Ztsi
+            {
+                ServiceUrl = desiredZtsi.DesiredServiceUrl
+            };
+            Twin twinPatch = CreateTwinPatch(ComponentName, desiredZtsi);
+
+            if (UpdateTwinBlockUntilUpdate(twinPatch))
+            {
+                Ztsi reportedObject = JsonSerializer.Deserialize<Ztsi>(GetNewTwin().Properties.Reported[ComponentName].ToString());
```

Comment:
Given we are testing reported properties only here, i would recommend not using `GetNewTwin()` as it performs an iot hub call for every call. Everything should be available in the `GetTwin()`. It will make your test faster and less-costly in the end